HuaPay Plugin Entities added

zvorak opened the pull request.

zvorak committed:
PaymentConfig entity added

zvorak committed:
PaymentMethodConfig entity added

zvorak committed:
each table migration, methods of PluginManager

zvorak commented on Entity/PaymentConfig.php:
```diff
@@ -0,0 +1,124 @@
+    /**
+     * @var string
+     */
+    private $api_token;
```
APIトークン。

zvorak commented on Entity/PaymentConfig.php:
```diff
@@ -0,0 +1,124 @@
+/**
+ * PaymentConfig
+ */
+class PaymentConfig extends \Eccube\Entity\AbstractEntity
```
決済情報を保持するエンティティ。PaymentMethodConfigと親子関係の構造を持ちながら外部決済サービスとの連携情報を保持する。
今後プラグイン機能の進展によって保持する情報が増えるケースに備えて、”huapay”のような広い意味を持ちうるエンティティ名ではなく、機能や役割を明示したエンティティ名を提唱するもの。

zvorak commented on Entity/PaymentConfig.php:
```diff
@@ -0,0 +1,124 @@
+    /**
+     * @var integer
+     */
+    private $is_testing;
```
テスト用設定(true)か、商用(false)か否か。

zvorak commented on Entity/PaymentConfig.php:
```diff
@@ -0,0 +1,124 @@
+    /**
+     * @var string
+     */
+    private $reference_prefix;
```
決済レコードを記録する際のプレフィックス。外部決済サーバーとECCUBEの間で紐付けの役割を持つ。

deraru replied:
これいらないかなーと私は思ったのですが、どう使われるのかとその必要性をもう少し説明していただけますか？

zvorak replied:
実際に不要かどうかは実運用に乗せてからでも良いかと思いますが如何でしょうか。
後から「やっぱり必要だった」となって付け足すよりは、いざという時に備えて用意だけしておくのであれば、カラム一つくらいはあっても良いかと、というご提案です。

deraru replied:
なるほどー。それであれば過剰設計ですのでカラムごと消してください。

zvorak replied:
はい、承知いたしました。
カラム削除は構いませんが、後々万一「やっぱり必要でした」となった場合の工数は、追加の稼働として対応させていただきますのでご了承くださいますようお願い申し上げます。

deraru replied:
了解です！
まあ必要になったとしても、カラムで保つのではなく定数で持たせる設計にすると思います。

zvorak commented on Entity/PaymentMethodConfig.php:
```diff
@@ -0,0 +1,180 @@
+/**
+ * PaymentMethodConfig
+ */
+class PaymentMethodConfig extends \Eccube\Entity\AbstractEntity
```
決済手段識別のためのエンティティ。AliPay、UnionPayなど決済手段ごとに1エンティティとなり、PaymentConfigエンティティと連携する。
親子関係を明示的にするため、エンティティ名はPaymentXxxConfigの形式を取るものとして推奨。

deraru replied:
> 親子関係を明示的にするため、エンティティ名はPaymentXxxConfigの形式を取るものとして推奨。

これPHPの流儀なんですかね？めちゃくちゃ違和感あるんですけども。
あとこれそもそも親子関係ではないですよね？

zvorak replied:
PHPは関係ありません。
PaymentConfigとの間で１対多の関係を作るのでそのように表現したものです。

deraru replied:
なるほどPHP的な親子関係ではなく、1対多の関係であるからこの名前なのですね。
であれば `HuapayPaymentMethod` としてください。本当は `PaymentMethod` で十分かとおもったのですが、EC-CUBE側で使われてそうな名前だったので変えました。
もしプラグインで名前空間があり、衝突の可能性がないのであれば `PaymentMethod` でお願いします。

zvorak replied:
テーブル名では自動生成の形で「plg_huapayplugin_xxx」(xxxはエンティティ名)として定義されており、「huapay」のようにプラグイン名と被る名称は重言ぽくなって馴染みにくいかと感じております。

deraru replied:
なるほど。では `PaymentMethod` でFixでお願いします。

https://github.com/huapay/huapay-ec-cube-plugin/pull/5#discussion_r214227399 に対する回答としては、Entity名を変更するので `name` カラムのままでいきましょう、となります。

zvorak commented on Entity/PaymentMethodConfig.php:
```diff
@@ -0,0 +1,180 @@
+    /**
+     * @var integer
+     */
+    private $payment_config_id;
```
親となるPaymentConfigエンティティのID。

zvorak commented on Entity/PaymentMethodConfig.php:
```diff
@@ -0,0 +1,180 @@
+    /**
+     * @var integer
+     */
+    private $eccube_payment_id;
```
ECCUBE本体側のPaymentエンティティID。本体側で決済情報を記録する際、このIDを使って記録されるため、本体とプラグインとの間のデータ紐付けの役割を持つ。

deraru replied:
ここちょっとよく分かっていないので、必要性をもう少し説明していただけますか？
各EntityのIDを使ってはだめなんでしょうか？
EC-CUBE側で払い出される `payment_id` があって、それをプラグイン側で持っておく必要がある、ということでしょうか？

zvorak replied:
> EC-CUBE側で払い出される payment_id があって、それをプラグイン側で持っておく必要がある、ということでしょうか？

はい、現時点では少なくともそのように把握しております。
「現時点では少なくとも」というのは、実装を進めるうちにプラグイン側でこのIDを保持しなくてもいいようにできる工夫の方法は否定致しませんが、具体的にその方針での実装案を持っておりませんので、この形で進めさせて欲しいということになります。

zvorak commented on Entity/PaymentMethodConfig.php:
```diff
@@ -0,0 +1,180 @@
+    /**
+     * @var string
+     */
+    private $name;
```
決済手段の名称(Alipay、UnionPayなど)。当エンティティにおいて他に名前(name)に相当する属性値が存在しないため、修飾語なしで単なる「名前(name)」とした方が良いのではないかとしてご提案いたします。

deraru replied:
エンティティ名が `PaymentMethodConfig` で、 `name` になると意味が分からなくならないですかね？
もし `name` にするならエンティティ名を変更したほうが良いと思います。
私としては `name` カラムを維持してエンティティ名を変更した方が良いのではないかと思いますがどうでしょうか？

zvorak replied:
決済機関情報を複数のエンティティに分けたデータ構造とする以上、IDだけで識別するのもデータ運用上直感的でなく戸惑うこともあるかと思いますのでhuman readableな形で名前を入れたい、という目的のものなので"name"以外に長い名前を持つカラムは直感的でないのではないかと。
それよりは、ここの名前で見解が分かれて難しい話になるようであれば、最初の単一エンティティ構造に戻してalipay/unionpay/wechatpayの各決済手段の情報も全部カラムとして持つ方がシンプルにできて良いかなとも思いますが。

zvorak commented on Entity/PaymentMethodConfig.php:
```diff
@@ -0,0 +1,180 @@
+    /**
+     * @var integer
+     */
+    private $enabled;
```
当該決済情報が有効化されているかどうか。(“is_enabled”に修正の予定)

zvorak commented on Entity/PaymentMethodConfig.php:
```diff
@@ -0,0 +1,180 @@
+    /**
+     * @var string
+     */
+    private $reference_prefix;
```
不要。削除予定。

zvorak committed:
enabled => is_enabled, reference_prefix

deraru reviewed (commented):
テーブルのNULL制約周りもみていてだけますか。ほとんどNOT NULLでよいはずなので。

zvorak commented:
テーブル定義はdcm.ymlファイルの方にございますのでそちらをレビューしましょう。

deraru commented:
レビューしてNOT NULLではなかったのでコメントしております。

zvorak commented on Resource/doctrine/Plugin.HuaPayPlugin.Entity.PaymentMethodConfig.dcm.yml:
```diff
@@ -0,0 +1,32 @@
+            unsigned: false
+        name:
+            type: text
+            nullable: true
```
NULL制約の件は具体的にこの辺りの記述を指してらっしゃいますよね、という話がしたかったので @deraru 
ファイル位置を指しながらコメント入れ、レビューをするご指定でしたのでそれに沿って進めさせていただければと、その確認です。

deraru replied:
当初設計と異なる箇所には、なぜ変えたのかとその必要性をコメントしていただけますか？実装上変えた、というのは全然受け入れますので。

zvorak commented on Resource/doctrine/Plugin.HuaPayPlugin.Entity.PaymentConfig.dcm.yml:
```diff
@@ -0,0 +1,22 @@
+        reference_prefix:
+            type: text
+            nullable: true
+    lifecycleCallbacks: {  }
```
上3行トリ

zvorak committed:
Entities renamed:
  PaymentConfig => Payment
  PaymentMethodConfig => PaymentMethod

Doctrine configurations and several attributes revised:
  nullability, default value, indexes, etc.

zvorak commented on Controller/ConfigController.php:
```diff
@@ -27,14 +27,22 @@ class ConfigController
     public function index(Application $app, Request $request)
     {
 
+	$PaymentRepository = $app['eccube.plugin.repository.payment'];
```
Plugin\HuaPayPlugin\Repository\PaymentRepository を使用してエンティティを取得する処理のモックアップ。index()メソッドを通して現時点では特に何かするものではなく、ID=1のエンティティを決め打ちで取得できることを確認するだけ。

zvorak commented on Entity/Payment.php:
```diff
@@ -0,0 +1,124 @@
+    /**
+     * @var string
+     */
+    private $api_token;
```
HuaPay API接続のためのAPIトークン

zvorak commented on Entity/Payment.php:
```diff
@@ -0,0 +1,124 @@
+    /**
+     * @var integer
+     */
+    private $is_testing;
```
テスト環境ならば1(正確には非0)、商用ならば0

zvorak commented on Entity/Payment.php:
```diff
@@ -0,0 +1,95 @@
+     * @var integer
+     */
+    private $is_testing;
+
```
ここより下はsetter/getter (特別な処理なし)

zvorak commented on Entity/PaymentMethod.php:
```diff
@@ -0,0 +1,152 @@
+    /**
+     * @var integer
+     */
+    private $payment_id;
```
ECCUBE本体側のPaymentエンティティID。

zvorak commented on Entity/PaymentMethod.php:
```diff
@@ -0,0 +1,152 @@
+    /**
+     * @var string
+     */
+    private $name;
```
決済手段の名称。AliPayやUnionPayなど、接続先となる外部の決済機関を識別する名前を持つ。

zvorak commented on Entity/PaymentMethod.php:
```diff
@@ -0,0 +1,152 @@
+    /**
+     * @var integer
+     */
+    private $is_enabled;
```
有効化された決済手段であれば1(非0)、無効化されて入れば0。

zvorak commented on PluginManager.php:
```diff
@@ -36,6 +36,9 @@ public function install($config, Application $app)
+        $this->disableAllPayment($app);
+
+        $this->migrationSchema($app, __DIR__.'/Resource/doctrine/migration', $config['code'], 0);
     }
```
プラグインをアンインストールする際の処理を記述。

- プラグインで設定した決済手段を削除
- プラグイン用に作成したテーブルを削除

zvorak commented on PluginManager.php:
```diff
@@ -47,6 +50,19 @@ public function uninstall($config, Application $app)
+            $em->getConnection()->rollback();
+            throw $e;
+	}
     }
```
プラグインを有効化した際の処理を記述。

- プラグインで使用する、エンティティ用のテーブルを生成。
- プラグインで使用する決済手段をECCUBE本体のDBにも登録(ただし現時点ではモックアップ) (58行目〜60行目までの間に実装する予定)

zvorak commented on PluginManager.php:
```diff
@@ -58,6 +74,7 @@ public function enable($config, Application $app)
     public function disable($config, Application $app)
     {
+        $this->disableAllPayment($app);
     }
```
プラグインを無効化する際の処理。
ECCUBE本体からプラグイン用の決済情報を削除するが、アンインストール時とは異なりプラグインエンティティ用のテーブルは削除しない。

zvorak commented on PluginManager.php:
```diff
@@ -71,4 +88,68 @@ public function update($config, Application $app)
     {
     }
 
```
以下、ECCUBE本体の決済情報操作のためのヘルパー的メソッド。enable()メソッドのモックアップ箇所等から呼び出される予定。

zvorak commented on ServiceProvider/HuaPayPluginServiceProvider.php:
```diff
@@ -38,6 +38,12 @@ public function register(BaseApplication $app)
+        });
+        $app['eccube.plugin.repository.paymentmethod'] = $app->share(function () use ($app) {
+            return $app['orm.em']->getRepository('Plugin\HuaPayPlugin\Entity\PaymentMethod');
+        });
```
エンティティ取得用のリポジトリクラスの登録。

deraru reviewed (approved):
LGTM!
